src/common/registry.rs

```diff
@@ -27,44 +27,112 @@ pub enum RegistryError {
     ModelNotFound(String),
     #[error("RegistryError : model '{0}' already exists")]
     ModelAlreadyRegistered(&'static str),
+    #[error("RegistryError: model '{0}' dependency getter already registered")]
+    ModelDependencyFuncRegistered(&'static str),
+    #[error("RegistryError: dependencies unsupported for component type '{0}'")]
+    ComponentTypeNotInDependencies(&'static str),
+    #[error("RegistryError: model '{0}' not found in dependencies under component type '{1}'")]
+    ModelNotFoundInDependencies(String, &'static str),
 }
 
 use std::collections::BTreeMap as Map;
 
+use crate::proto::common::v1::ResourceName;
+
 use super::{
     board::BoardType, config::ConfigType, encoder::EncoderType, motor::MotorType,
-    movement_sensor::MovementSensorType, sensor::SensorType,
+    movement_sensor::MovementSensorType, robot::Resource, sensor::SensorType,
 };
 
-type MotorConstructor = dyn Fn(ConfigType, Option<BoardType>) -> anyhow::Result<MotorType>;
+pub fn get_board_from_dependencies(deps: Vec<Dependency>) -> Option<BoardType> {
+    for Dependency(_, dep) in deps {
+        match dep {
+            Resource::Board(b) => return Some(b.clone()),
+            _ => continue,
+        }
+    }
+    None
+}
 
-type SensorConstructor = dyn Fn(ConfigType, Option<BoardType>) -> anyhow::Result<SensorType>;
+// ResourceKey is an identifier for a component to be registered to a robot. The
+// first element is a string representing the component type (arm, motor, etc.)
+// and the second element is its name.
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+pub struct ResourceKey(pub &'static str, pub String);
+
+impl ResourceKey {
+    pub fn new(model: &str, name: String) -> Result<Self, anyhow::Error> {
+        let model_str = match model {
+            "motor" => crate::common::motor::COMPONENT_NAME,
+            "board" => crate::common::board::COMPONENT_NAME,
+            "encoder" => crate::common::encoder::COMPONENT_NAME,
+            "movement_sensor" => crate::common::movement_sensor::COMPONENT_NAME,
+            "sensor" => crate::common::sensor::COMPONENT_NAME,
+            &_ => {
+                anyhow::bail!("component type {} is not supported yet", model.to_string());
+            }
+        };
+        Ok(Self(model_str, name))
+    }
+}
 
-type MovementSensorConstructor =
-    dyn Fn(ConfigType, Option<BoardType>) -> anyhow::Result<MovementSensorType>;
+impl TryFrom<ResourceName> for ResourceKey {
+    type Error = anyhow::Error;
+    fn try_from(value: ResourceName) -> Result<Self, Self::Error> {
+        let comp_type: &str = &value.subtype;
+        let comp_name = match comp_type {
+            "motor" => crate::common::motor::COMPONENT_NAME,
```

Comment:
Is Board missing on purpose?

Reply:
since we always add the board first, I didn't end up needing it. But I can add it